multihash_derive/src/lib.rs

```diff
@@ -0,0 +1,229 @@
+                        match self {
+                            #(#code_matches_into_str)*
+                        }
+                    }
+                }
+
+                impl Code {
+                    pub fn from_u32(val: u32) -> Option<Self> {
+                        match val {
+                            #(#code_matches_from_u32)*
+                            _ => None
+                        }
+                    }
+                //     fn new_hasher(&self) -> Box<MultihashDigest> {
+                //         match *self {
+                //             Code::Sha1 => {
+                //                 Box::new(Sha1::new())
+                //             },
+                //             Code::Sha2256 => {
+                //                 Box::new(Sha2256::new())
+                //             },
+                //             _ => unimplemented!(),
+                //         }
+                //     }
+                }
+
+                #(#impls)*
+            };
+
+            res
+        }
+        _ => panic!("Multihash can only be derived on an enum"),
+    }
+}
+
+fn impl_variant(_code_name: &syn::Ident, variant: &syn::Variant) -> proc_macro2::TokenStream {
+    let name = &variant.ident;
+    let attrs = &variant.attrs;
+
+    let size: usize = fetch_attr("Size", &attrs)
+        .expect("Please supply a Size attribute")
+        .parse()
+        .expect("Size should be a number");
+
+    if size > 255 {
+        panic!("Sizes larger than 255 are not yet supported");
+    }
+
+    let code =
+        fetch_discriminant(&variant.discriminant).expect("Please provide the code as discriminant");
+
+    let mut code_varint = Vec::with_capacity(16);
+    code_varint.write_varint(code).unwrap();
+    let code_len = code_varint.len();
+    let code_iter = code_varint
+        .iter()
+        .enumerate()
+        .map(|(i, v)| quote!(out[#i] = #v;));
+
+    let original_digest: syn::Path =
+        syn::parse_str(&fetch_attr("Digest", &attrs).expect("Please provide a Digest"))
+            .expect("invalid digest provided");
+
+    // TODO: actually derive this from the encoded size of the size.
+    let prefix_len = 1 + code_len;
+
+    quote!{
+        #[derive(Debug, Default, Clone)]
+        pub struct #name(#original_digest);
+
+        impl Reset for #name {
+            fn reset(&mut self) {
+                digest::Digest::reset(&mut self.0);
+            }
+        }
+
+        impl Input for #name {
+            fn input<B: AsRef<[u8]>>(&mut self, data: B) {
+                digest::Digest::input(&mut self.0, data);
+            }
+        }
+
+        impl BlockInput for #name {
+            type BlockSize = <#original_digest as BlockInput>::BlockSize;
+        }
+
+        impl MultihashDigest for #name {
+            fn new() -> Self {
+                #name(<#original_digest as Digest>::new())
+            }
+
+            fn size() -> usize {
```

Comment:
Honest question: Why does this function return `usize` instead of `u8`? I see that you're doing some runtime checks to ensure that all `#size` values are in the range of unsigned, 8-bit integers anyways.

Reply:
Size could be larger, (I probably should change it to a `u32`), and while it is currently restricted, I want to fix that later. By defining the signature here to be larger, I can add that later, without breaking any dependent code.

Reply:
fixed to u23